control: Implement "Final" Version of High Level Control

mrrosen opened the pull request:
Flesh our high level control FSM more and some bug fixes associated with its development. Provide framework for interfacing with mid and low level controllers.

Commit:
control: flesh out control fsm

Further refine high-level controller and make clearer the separation
between high- and mid-level control loops

Signed-off-by: Rosen, Michael R <michael.r.rosen@intel.com>

JonnerLynx commented on src/modules/control.cpp:
```diff
@@ -2,60 +2,117 @@
-      }
+      // Collect all set points from parameters
+      totalBreathTime = (60 SEC) / parameters.respirationRateRequested;
+      targetInhalationTime = (parameters.ieRatioRequested * totalBreathTime) >> 8;
```
The right shift by 8 is a little confusing here. I'd recommend either adding a comment explaining what it does, or finding a different way to express this.

mrrosen replied:
Yeah, will need to comment how what ieRatio is/type

JonnerLynx replied:
Yeah, let's place the details of the fixed-point multiplication inside of a macro or a static inline function.

JonnerLynx commented on src/modules/control.cpp:
```diff
@@ -2,60 +2,117 @@
+      
+      // Kickoff timers for monitoring breath/stages of breathing
+      timerHalBegin(&breathTimer, totalBreathTime);
+      timerHalBegin(&controlTimer, targetInhalationTime * INHALATION_OVERTIME);
```
I think that the inspiratory pause needs to be included as part of the inhalation time in order to respect the I:E ratio. Also, I'm not sure if it makes sense for the length of the pause to be dependent upon the total inhalation time.

JonnerLynx replied:
See comment above regarding `INHALATION_OVERTIME`.

JonnerLynx committed:
Refactored motor driver for glitch-free PWM output.

JonnerLynx committed:
Updated motor driver with speed control. Moved motor state control out of ISR.

JonnerLynx commented on src/modules/control.cpp:
```diff
@@ -2,60 +2,117 @@
+#include "../modules/sensors.h"
 #include "../modules/parameters.h"
 
+#define INHALATION_OVERTIME 4 / 3
```
This definition is a bit fragile in that it will not work correctly with enclosing parenthesis, i.e. `targetInhalationTime*(INHALATION_OVERTIME)`

I'd suggest turning this into a macro or a static inline function that forces the correct order of arithmetic operator evaluation.

JonnerLynx commented on src/modules/control.cpp:
```diff
@@ -2,60 +2,117 @@
 #include "../modules/parameters.h"
 
+#define INHALATION_OVERTIME 4 / 3
+#define HOLD_TIME           (150 MSEC)
```
The guidance from Saurabh is to employ an inspiratory pause of 200 ms. Are you open to changing the name of this definition to something like INSPIRATORY_PAUSE_TIME?

mrrosen replied:
Yes, we can certainly make this 200 instead

JonnerLynx commented on src/modules/control.cpp:
```diff
@@ -2,60 +2,117 @@
+      uint16_t motorDuration = 0;
+  
+      // If in a volume controlled mode, run the Mid-level volume controller
+      if ((parameters.ventilationMode == VENTILATOR_MODE_VC) ||
+          (parameters.ventilationMode == VENTILATOR_MODE_AC)) {
```
nit: `ventilationMode` vs. `VENTILATOR_MODE_*`

Consider switching definitions to `VENTILATION_MODE_*`.

mrrosen replied:
Yeah, modes need to be cleaned up anyway -- still waiting to hear back from Saurabh about it

JonnerLynx commented on src/modules/control.cpp:
```diff
@@ -65,14 +122,33 @@ static PT_THREAD(controlThreadMain(struct pt* pt))
-      // TODO: Fix this condition for assisted modes
-      PT_WAIT_UNTIL(pt, motorHalRun() != HAL_IN_PROGRESS && timerHalRun(&controlTimer) != HAL_IN_PROGRESS);
+      // Wait for the motor to move back
+      // TODO: consider if patient takes breathe before motor has completely moved back?
```
nit: breathe -> breath

JonnerLynx commented on src/modules/control.cpp:
```diff
@@ -65,14 +122,33 @@ static PT_THREAD(controlThreadMain(struct pt* pt))
-      motorHalBegin(MOTOR_HAL_DIRECTION_EXHALATION, motorCompressionDistance, motorCompressionDuration);
-      timerHalBegin(&controlTimer, 1 MSEC); // TODO: calculate this value from the breathe parameters
+      // TODO: Andrew, thoughts?; also, if this is a motor home, maybe need different HAL function for it
+      motorHalBegin(MOTOR_HAL_DIRECTION_EXHALATION,
+                    totalMotorCompressionDistance,
+                    totalMotorCompressionDuration);
```
I don't think that it's necessary to use totoalMotorCompressionDuration here, but I think it's a moot point as we are moving to a different method anyways (speed control instead of duration command).

JonnerLynx commented on src/modules/control.cpp:
```diff
@@ -65,14 +122,33 @@ static PT_THREAD(controlThreadMain(struct pt* pt))
+      // Calculate and update the measured times
+      uint32_t currentBreathTime = timerHalCurrent(&breathTimer);
+      measuredExhalationTime = currentBreathTime - HOLD_TIME - measuredInhalationTime;
+      control.ieRatioMeasured = (measuredExhalationTime << 8) / measuredInhalationTime;
```
Let's do all fixed point math operations inside a macro or a static inline function.

JonnerLynx reviewed (approved):
I like how the high-level control is evolving here. I raised a few issues, but nothing worth blocking the merge for. Please make any updates that you would like to and merge to develop.

arbarkan committed:
Add open-loop trajectory tracking control.

jast1982 committed:
First implementation of closed loop PID control

not tested in HW yet

zmk5 commented on src/modules/control.cpp:
```diff
@@ -2,60 +2,117 @@
+      totalMotorCompressionDuration += motorDuration;
+      
+      // If controller reports nonzero distance, need to move motor
+        if (motorDistance != 0) {
+          
+        }
```
This may be pedantic but the indentation is a bit off.

Commit:
control: use metrics to time control loop

Signed-off-by: Rosen, Michael R <michael.r.rosen@intel.com>

mrrosen commented:
Heavily refactored to merge in midlevel control constructs

Commit:
control: temporarilly fix requested IE ratio

Currently the serial control sends the IE as an enumeration when
it should be a fixed point number; for now fix the value to 1:1.5

Signed-off-by: Rosen, Michael R <michael.r.rosen@intel.com>

Commit:
control: fix a number of issues from Jan

. Fix minor issues in stop mid control loop
. Refactor timers to be periodic to improve precision of timing loops

Signed-off-by: Rosen, Michael R <michael.r.rosen@intel.com>

Commit:
sensors: refactor sensors hal and module significantly

Using Jan's work as a base, refactor sensor hal to sample ADC
without timer. Refactor alot of the sensor module to incorporate
the refactored sensors hal and timer hal.

Signed-off-by: Rosen, Michael R <michael.r.rosen@intel.com>

Commit:
sensors: add safety yields to sensor threads

Signed-off-by: Rosen, Michael R <michael.r.rosen@intel.com>

jast1982 committed:
add initial controller scaling

added scaling factor to trajectory and motor output according to Andrews
measurements.
reset I-integrator for every cycle

Commit:
debug: add debug plotter mechanism

In order to take advantage of the Arduino Serial Plotter, debug
messages mechanism had to be slightly retuned and now the debug
feature includes a way to plot values with some limitations.

Signed-off-by: Rosen, Michael R <michael.r.rosen@intel.com>

jast1982 committed:
First controller tuning...

Dirty fix to motor target position
initial controller tuning - works, but looses steps and can't compensate
(yet)

Commit:
control: minor fixes to control module

. Add macro to enable/disable closed loop control while under
  development
. Fix typo in timeout check for exhalation

Signed-off-by: Rosen, Michael R <michael.r.rosen@intel.com>

jast1982 committed:
changed motor driver to speed only control

changed motor driver to speed only interface
added some sanity checks for inputs

jast1982 committed:
mid-level controller update

- new trajectory generation
- automatic volume adjustment active
- automatic pressure adjustment implemented, but inactive
- new output mapping to adjust for angle dependency in air flow speed
- new, lot simpler and faster exhale control trajectory
- continouus motor output filter also between inhale/exhale state switch
- all parameters are used from GUI except I:E ratio
- flow controller tuned again

Commit:
sensors: increase pressure threshold for inhalation detection

Inhalation detection will likely need to be heavily reworked,
for now, the inhalation threshold was only 1cmH2O, but was too
low and triggering incorrectly; not its at 8cmH2O

Signed-off-by: Rosen, Michael R <michael.r.rosen@intel.com>

Commit:
alarm: fix alarms to use the new timer api

Signed-off-by: Rosen, Michael R <michael.r.rosen@intel.com>
